library/tract-renderer/src/main/java/org/cru/godtools/tract/activity/TractActivity.java

```diff
@@ -335,34 +375,74 @@ private boolean validStartState() {
-            mDownloadManager.cacheTranslation(mTool, mLanguages[mActiveLanguage]);
-            mDownloadTask = SettableFuture.create();
-            AsyncTask.THREAD_POOL_EXECUTOR
-                    .execute(new DownloadTranslationRunnable(this, mTool, mLanguages[mActiveLanguage], mDownloadTask));
+    private void downloadTranslations() {
+        if (mDownloadManager != null && mTool != null) {
+            for (int language = 0; language < mLanguages.length; language++) {
+                mDownloadManager.cacheTranslation(mTool, mLanguages[language]);
+                mDownloadTasks[language] = SettableFuture.create();
+                AsyncTask.THREAD_POOL_EXECUTOR.execute(
+                        new DownloadTranslationRunnable(this, mTool, mLanguages[language], mDownloadTasks[language]));
+            }
         }
     }
 
-    private void updateVisibilityState() {
-        final int visibilityState;
-        if (getActiveManifest() != null) {
-            visibilityState = STATE_ACTIVE;
-        } else if (mDownloadTask != null && mDownloadTask.isDone() &&
-                mTranslations.indexOfKey(mActiveLanguage) >= 0 && mTranslations.get(mActiveLanguage) == null) {
-            visibilityState = STATE_NOT_FOUND;
+    private int determineLanguageState(final int languageIndex) {
+        if (mManifests.get(languageIndex) != null) {
+            return STATE_LOADED;
+        } else if (mDownloadTasks[languageIndex] != null && mDownloadTasks[languageIndex].isDone() &&
+                mTranslations.indexOfKey(languageIndex) >= 0 && mTranslations.get(languageIndex) == null) {
+            return STATE_NOT_FOUND;
         } else {
-            visibilityState = STATE_LOADING;
+            return STATE_LOADING;
         }
+    }
+
+    /**
+     * This method updates the list of artificially hidden languages. This includes primary language fallbacks provided
+     * via a deep link.
+     */
+    @UiThread
+    @VisibleForTesting
+    void updateHiddenLanguages() {
+        int primaryLanguage = -1;
+        for (int i = 0; i < mLanguages.length; i++) {
+            // default hidden state to whether the language exists or not
+            final int state = determineLanguageState(i);
+            mHiddenLanguages[i] = state == STATE_NOT_FOUND;
+
+            // primary language specific logic
+            if (i < mPrimaryLanguages && state != STATE_NOT_FOUND) {
+                if (mActiveLanguage == i || (state == STATE_LOADED && primaryLanguage == -1)) {
+                    // don't hide the primary language
+                    mHiddenLanguages[i] = false;
+
+                    // hide any previously identified primary languages
+                    if (primaryLanguage != -1) {
+                        mHiddenLanguages[primaryLanguage] = true;
+                    }
+
+                    // track our current primary language
+                    primaryLanguage = i;
+                } else if (i < mPrimaryLanguages) {
```

Comment:
How could this block get called? `i >= mPrimaryLanguages` would be gated on line 414.

Reply:
Never mind.. sleep deprivation